cirq/protocols/apply_unitary.py

```diff
@@ -227,27 +264,48 @@ def apply_unitary(unitary_value: Any,
     if func is not None:
         result = func(args)
         if result is not NotImplemented and result is not None:
+            if result is args.target_tensor:
```

Comment:
This is not sufficient. For example, it could fail if a user makes the mistake of saying `args.target_tensor *= -0.5`, which is something that HPowGate does.

I think what we need to do is slice down the tensor to the shape expected by the gate before passing it along, in order to guarantee that it is working in a subspace that it understands. Then we can separately handle the remainder using this approach.

Reply:
Fixed with new args slicing method.